Lesson23 page factory

ochaklosh opened the pull request:
DO NOT MERGE THIS WITH MAIN
It is an unsuccessful try of using pageFactory, since the Login method is no longer working
During the instructional period the code was not working as well.

Updated HomePage - moved locator information withing a method to the Locators section
Went back to Homework22Tests branch and edited, which I will merge with main branch

ochaklosh committed:
Update ActionsTest.java

Still working

ochaklosh committed:
Update BasePage.java

Added method isPlaying()

ochaklosh committed:
Update HomePage.java

Still working

ochaklosh committed:
Update TestNG.xml

Updated which test to run

ochaklosh committed:
Update LoginTests.java

Currently with using PageFactoring Login does not work it gives mistake
Cannot invoke "org.openqa.selenium.WebElement.clear()" because "this.emailField" is null
which is a method to clear the emailField before entering users email

ochaklosh committed:
Create BasePage.java

Created BasePage to use with PageFactory

ochaklosh committed:
Created LoginPage

Created LoginPage to use pageFactory, so far unsuccessfully

ochaklosh committed:
Update AllSongsPage.java

Updated the find method from
findElementAndClickability   to
 waitForVisibility

ochaklosh committed:
Update HomePage.java

Updated by removing locator information within a method and add to the Locator section

ochaklosh committed:
Update TestNG.xml

Updated which test I want to run

Berichicko commented:
Hi dear Oksana

I'll help you again -)

![image](https://github.com/ochaklosh/A59-QA-Automation-Framework/assets/104761482/57e3cc63-1c7f-4345-b956-3bd9c6b65199)

![image](https://github.com/ochaklosh/A59-QA-Automation-Framework/assets/104761482/b5bb697b-8983-443d-a75e-bfefce8e41be)

**HINT**

**Step 1**

```
package pageFactory;

import org.openqa.selenium.*;
import org.openqa.selenium.interactions.Actions;
import org.openqa.selenium.support.PageFactory;
import org.openqa.selenium.support.ui.ExpectedConditions;
import org.openqa.selenium.support.ui.WebDriverWait;
import java.time.Duration;


public class BasePage {

    protected WebDriver driver;
    protected WebDriverWait wait;
    protected Actions actions;

    public BasePage(WebDriver givenDriver) {
        driver = givenDriver;
        wait = new WebDriverWait(driver, Duration.ofSeconds(10));
        actions = new Actions(driver);
        PageFactory.initElements(driver, this);
    }

    public WebElement waitForVisibility(WebElement element) {
        return wait.until(ExpectedConditions.visibilityOf(element));
    }

     public WebElement findElementAndClickability(WebElement element) {
        return wait.until(ExpectedConditions.elementToBeClickable(element));
    }

}
```

**Step 2**

```
package pageFactory;

import org.openqa.selenium.By;
import org.openqa.selenium.WebDriver;
import org.openqa.selenium.WebElement;
import org.openqa.selenium.support.FindBy;


public class LoginPage extends BasePage {


    public LoginPage(WebDriver givenDriver){
        super(givenDriver);
    }


    @FindBy(css = "input[type='email']")
    private WebElement emailField;

    @FindBy(css = "input[type='password']")
    WebElement passwordField;

    @FindBy(css = "button[type='submit']")
    WebElement submitBtn;

    //PAGE METHODS
    public LoginPage provideEmail(String email){
        emailField.clear();
        waitForVisibility(emailField).sendKeys(email);
          return this;
    }
    public LoginPage providePassword(String password){
        passwordField.clear();
        waitForVisibility(passwordField).sendKeys(password);
        return this;
    }
    public LoginPage clickSubmit(){
        findElementAndClickability(submitBtn).click();
        return this;
    }

    public void login(){
        provideEmail("demo@testpro.io");
        providePassword("te$t$tudent");
        clickSubmit();
    }
}
```

**Step 3**

```
package pageFactory;

import org.openqa.selenium.WebDriver;
import org.openqa.selenium.WebElement;
import org.openqa.selenium.support.FindBy;

public class HomePage extends BasePage{


  public HomePage(WebDriver givenDriver){
    super(givenDriver);
  }


  @FindBy(css = "img.avatar")
  private WebElement userAvatarIcon;


  public WebElement getUserAvatar(){
    return waitForVisibility(userAvatarIcon);
  }

}
```

**Step 4**

```
import org.apache.commons.logging.LogFactory;
import org.testng.Assert;
import org.testng.annotations.Test;

import pageFactory.HomePage;
import pageFactory.LoginPage;
public class LoginTests extends BaseTest {


    private static final Log log = LogFactory.getLog(LoginTests.class);


    //Happy Path - Login Test
    @Test
    public void loginValidEmailValidPassword(){

        LoginPage loginPage = new LoginPage(driver);
        HomePage homePage = new HomePage(driver);

        loginPage.login();
         Assert.assertTrue(homePage.getUserAvatar().isDisplayed());


    }



    @Test(dataProvider = "NegativeLoginTestData", dataProviderClass = TestDataProvider.class)
    public void negativeLoginTests(String email, String password){
        LoginPage loginPage = new LoginPage(driver);
        loginPage.login();
        Assert.assertEquals(driver.getCurrentUrl(), url);

    }
}
```

![image](https://github.com/ochaklosh/A59-QA-Automation-Framework/assets/104761482/4b9122ac-ca69-48af-8282-9a70b74e999d)

![image](https://github.com/ochaklosh/A59-QA-Automation-Framework/assets/104761482/d9a671ef-fe12-4bd5-80bf-e2b26e7572b7)

**INFO:** In your code, the methods have different parameter types: one method takes a WebElement and another takes a By. The ClassCastException error happens because you are trying to cast a WebElement to a By, which is not possible.

For example

Method with **By Parameter** (POM)

![image](https://github.com/ochaklosh/A59-QA-Automation-Framework/assets/104761482/e7e4e5b5-a8df-413d-9369-fae00952201d)

Method with **WebElement Parameter** (Factory)

![image](https://github.com/ochaklosh/A59-QA-Automation-Framework/assets/104761482/c4d32ba9-da9a-46a0-85d4-7242299ecc2e)

**Page Object Model and Page Factory**

![POM_vs_Factory_1](https://user-images.githubusercontent.com/104761482/236864030-34b89652-460d-4440-aa91-894332e90654.png)

![POM_vs_Factory_2](https://user-images.githubusercontent.com/104761482/236864098-d3ffe636-7ba9-484e-a72f-8b03ff024659.png)

ochaklosh committed:
Update BasePage.java

Updated waitForVisibilityPF and findElementAndClickability methods

to work with pageFactory t provide WebElement instead of a locator

ochaklosh committed:
Create HomePage.java

Added HomePage.java to pageFactory package
Updated locator userAvatarIcon to be used with @FindBy

ochaklosh committed:
Update LoginPage.java

Updated methods and imports to support PageFactory
